ChangeProjectVersion should ignore Maven CI-friendly version placeholders

`ChangeProjectVersion` implementation in the rewrite-maven module should ignore Maven CI-friendly version placeholder properties such as `${revision}`, `${sha1}`, and `${changelist}` similar to Implicitly Defined Version Properties.

**Use Case**

Automate the use of [Maven CI Friendly versioning](https://maven.apache.org/guides/mini/guide-maven-ci-friendly.html) within maven apps. 

To enable this automation, custom declarative recipes are created and packaged as jar for consumption.

**Workflow and Recipe Definition**

- Add `revision` property using `org.openrewrite.maven.AddProperty`

```xml
mvn org.openrewrite.maven:rewrite-maven-plugin::RELEASE:run  -Drewrite.activeRecipes=org.openrewrite.maven.AddProperty -Drewrite.options=key=revision,value=${VERSION}
```

- Add custom declarative recipe to change project version to `${revision}`.

```
---
type: specs.openrewrite.org/v1beta/recipe
name: customRecipe
displayName: Change Project Version to Use Revision Property
description: >
  Updates the project version to use the ${revision} property for CI friendly versioning.
recipeList:
  - org.openrewrite.maven.ChangeProjectVersion:
      groupId: '*'
      artifactId: '*'
      newVersion: ${revision}
```
This recipe is executed via maven command line with rewrite option `-Drewrite.resolvePropertiesInYaml=false` to avoid interpolation of `revision` value

When the custom declarative recipe to change the project version (e.g., setting it to ${revision} for CI-friendly versioning), the recipe works as expected on the first run. However, subsequent runs of the recipe override the `revision` property with ${revision}  again, leading to a cyclic reference error.

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
<!--
Whenever possible please try to replicate your issue with the latest versions of OpenRewrite.
The latest major and minor versions of OpenRewrite projects are usually listed here:
https://docs.openrewrite.org/reference/latest-versions-of-every-openrewrite-module
For patch releases check the GitHub Releases page for the respective project.

We release every few weeks, so it's possible that your issue has already been fixed.

If you want to try the most recent changes that haven't been fully released yet, you can check out our snapshot releases
https://docs.openrewrite.org/reference/snapshot-instructions
-->
I am using

- Maven plugin v6.11.0
- rewrite-module v8.55.x

## How are you running OpenRewrite?
<!--
Are you using the Maven plugin, Gradle plugin, Moderne CLI, Moderne.io or something else?
Is your project a single module or a multi-module project?

Can you share your configuration so that we can rule out any configuration issues?

Is your project public? If so, can you share a link to it?
Code snippets can also be shared privately via [our public Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA).
-->
I am using maven command line , and work with both single/multi module pojects

```xml
mvn org.openrewrite.maven:rewrite-maven-plugin::6.11.0:run -Drewrite.recipeArtifactCoordinates=custom-decalartive-recipes:RELESE:run-Drewrite.resolvePropertiesInYaml=false -Drewrite.activeRecipes=custom-declarative-recipe
```

## What is the smallest, simplest way to reproduce the problem?
<!--
Sometimes the logs indicate a recipe stumbled over a particular pattern of code.
If you can share a code snippet that reproduces the issue, that will help us fix it faster.
We also accept [pull requests that merely replicate an issue](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes), as a step up to a full fix.

A code snippet can be something simple like this, or similar for other languages:
-->

**Minimal Unit test**

```java

package com.example;

import org.junit.jupiter.api.Test;
import org.openrewrite.DocumentExample;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.maven.Assertions.pomXml;

public class OpenRewriteChangeProjectVersionTest implements RewriteTest {

    @DocumentExample
    @Test
    void doNotchangeProjectVersionIfRevisionPlaceholderProperty() {
        rewriteRun(
            spec -> spec.recipe(new org.openrewrite.maven.ChangeProjectVersion(
                "org.openrewrite", "rewrite-maven", "${revision}", null)),
            pomXml(
                "<project>\n"
                + "    <groupId>org.openrewrite</groupId>\n"
                + "    <artifactId>rewrite-maven</artifactId>\n"
                + "    <version>${revision}</version>\n"
                + "    <properties>\n"
                + "        <revision>8.55.5</revision>\n"
                + "    </properties>\n"
                + "</project>\n"
            )
        );
    }
}

```

## What did you expect to see?
<!-- A code snippet, or a description of the behavior you expected helps us write a test to ensure the issue is fixed. -->
The recipe should not attempt to update the version if it is already set to one of these placeholders.

## What did you see instead?
<!-- A code snippet, or a description of the behavior you saw instead of the above expected result. -->
Recipe updates the revision property with placeholder property.

## What is the full stack trace of any errors you encountered?
<!-- When errors occur, please include the output of `--stacktrace` for Gradle or `--debug` for Maven. -->
Stack Trace from Test failure
```
,doNotchangeProjectVersionIfRevisionPlaceholderProperty(com.example.OpenRewriteChangeProjectVersionTest)
%TRACES 
java.lang.AssertionError: Failed to run recipe at Cursor{Document->root}
        at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:639)
        at org.openrewrite.scheduling.RecipeRunCycle.handleError(RecipeRunCycle.java:289)
        at org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:229)
        at org.openrewrite.scheduling.RecipeStack.reduce(RecipeStack.java:60)
        at org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$9(RecipeRunCycle.java:178)
        at org.openrewrite.internal.InMemoryLargeSourceSet.lambda$edit$0(InMemoryLargeSourceSet.java:83)
        at org.openrewrite.internal.ListUtils.map(ListUtils.java:244)
        at org.openrewrite.internal.ListUtils.map(ListUtils.java:267)
        at org.openrewrite.internal.InMemoryLargeSourceSet.edit(InMemoryLargeSourceSet.java:82)
        at org.openrewrite.scheduling.RecipeRunCycle.editSources(RecipeRunCycle.java:177)
        at org.openrewrite.RecipeScheduler.runRecipeCycles(RecipeScheduler.java:84)
        at org.openrewrite.RecipeScheduler.scheduleRun(RecipeScheduler.java:41)
        at org.openrewrite.Recipe.run(Recipe.java:441)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:377)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
        at com.exmaple.OpenRewriteChangeProjectVersionTest.doNotchangeProjectVersionIfRevisionPlaceholderProperty(OpenRewriteChangeProjectVersionTest.java:35)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: org.openrewrite.internal.RecipeRunException: java.lang.IllegalArgumentException: Circular placeholder reference 'revision' in property definitions
        at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:281)
        at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:268)
        at org.openrewrite.Preconditions$Check.visit(Preconditions.java:166)
        at org.openrewrite.Preconditions$Check.visit(Preconditions.java:145)
        at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:268)
        at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:154)
        at org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$7(RecipeRunCycle.java:209)
        at io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:147)
        at org.openrewrite.table.RecipeRunStats.recordEdit(RecipeRunStats.java:67)
        at org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$8(RecipeRunCycle.java:205)
        ... 16 more
Caused by: java.lang.IllegalArgumentException: Circular placeholder reference 'revision' in property definitions
        at org.openrewrite.internal.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:94)
        at org.openrewrite.internal.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:115)
        at org.openrewrite.internal.PropertyPlaceholderHelper.replacePlaceholders(PropertyPlaceholderHelper.java:74)
        at org.openrewrite.maven.internal.MavenPomDownloader.projectPomsByGav(MavenPomDownloader.java:179)
        at org.openrewrite.maven.internal.MavenPomDownloader.<init>(MavenPomDownloader.java:137)
        at org.openrewrite.maven.internal.MavenPomDownloader.<init>(MavenPomDownloader.java:102)
        at org.openrewrite.maven.UpdateMavenModel.updateResult(UpdateMavenModel.java:163)
        at org.openrewrite.maven.UpdateMavenModel.visitDocument(UpdateMavenModel.java:136)
        at org.openrewrite.xml.tree.Xml$Document.acceptXml(Xml.java:149)
        at org.openrewrite.xml.tree.Xml.accept(Xml.java:53)
        at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:245)
        ... 25 more

%TRACEE 
```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->
Yes

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/ChangeProjectVersion.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/ChangeProjectVersion.java
@@ -38,7 +38,8 @@ import static org.openrewrite.internal.StringUtils.matchesGlob;
 public class ChangeProjectVersion extends Recipe {
     // there are several implicitly defined version properties that we should never attempt to update
     private static final Collection<String> implicitlyDefinedVersionProperties = Arrays.asList(
-            "${version}", "${project.version}", "${pom.version}", "${project.parent.version}"
+            "${version}", "${project.version}", "${pom.version}", "${project.parent.version}",
+            "${revision}", "${sha1}", "${changelist}"
     );
 
     @Option(displayName = "Group",
@@ -97,13 +98,16 @@ public class ChangeProjectVersion extends Recipe {
                             String oldVersion = resolvedPom.getValue(versionTagValue);
                             assert oldVersion != null;
 
-                            if (!oldVersion.equals(newVersion)) {
-                                if (versionTagValue.startsWith("${") && !implicitlyDefinedVersionProperties.contains(versionTagValue)) {
-                                    doAfterVisit(new ChangePropertyValue(versionTagValue.substring(2, versionTagValue.length() - 1), newVersion, false, false).getVisitor());
-                                } else {
-                                    doAfterVisit(new ChangeTagValueVisitor<>(versionTag.get(), newVersion));
+                            // Skip if the current version tag value is already equal to the new version
+                            if (!versionTagValue.equals(newVersion)) {
+                                if (!oldVersion.equals(newVersion)) {
+                                    if (versionTagValue.startsWith("${") && !implicitlyDefinedVersionProperties.contains(versionTagValue)) {
+                                        doAfterVisit(new ChangePropertyValue(versionTagValue.substring(2, versionTagValue.length() - 1), newVersion, false, false).getVisitor());
+                                    } else {
+                                        doAfterVisit(new ChangeTagValueVisitor<>(versionTag.get(), newVersion));
+                                    }
+                                    maybeUpdateModel();
                                 }
-                                maybeUpdateModel();
                             }
                         } else if (Boolean.TRUE.equals(overrideParentVersion)) {
                             // if the version is not present and the override parent version is set,

--- a/rewrite-maven/src/test/java/org/openrewrite/maven/ChangeProjectVersionTest.java
+++ b/rewrite-maven/src/test/java/org/openrewrite/maven/ChangeProjectVersionTest.java
@@ -303,4 +303,61 @@ class ChangeProjectVersionTest implements RewriteTest {
         );
     }
 
+    @Test
+    void doNotChangeProjectVersionIfRevisionPlaceholderProperty() {
+        rewriteRun(
+          spec -> spec.recipe(new ChangeProjectVersion("org.openrewrite", "rewrite-maven", "${revision}", null)),
+          pomXml(
+            """
+              <project>
+                  <groupId>org.openrewrite</groupId>
+                  <artifactId>rewrite-maven</artifactId>
+                  <version>${revision}</version>
+                  <properties>
+                      <revision>8.55.5</revision>
+                  </properties>
+              </project>
+              """
+          )
+        );
+    }
+
+    @Test
+    void doNotChangeProjectVersionIfSha1PlaceholderProperty() {
+        rewriteRun(
+          spec -> spec.recipe(new ChangeProjectVersion("org.openrewrite", "rewrite-maven", "${sha1}", null)),
+          pomXml(
+            """
+              <project>
+                  <groupId>org.openrewrite</groupId>
+                  <artifactId>rewrite-maven</artifactId>
+                  <version>${sha1}</version>
+                  <properties>
+                      <sha1>abc123</sha1>
+                  </properties>
+              </project>
+              """
+          )
+        );
+    }
+
+    @Test
+    void doNotChangeProjectVersionIfChangelistPlaceholderProperty() {
+        rewriteRun(
+          spec -> spec.recipe(new ChangeProjectVersion("org.openrewrite", "rewrite-maven", "${changelist}", null)),
+          pomXml(
+            """
+              <project>
+                  <groupId>org.openrewrite</groupId>
+                  <artifactId>rewrite-maven</artifactId>
+                  <version>${changelist}</version>
+                  <properties>
+                      <changelist>SNAPSHOT</changelist>
+                  </properties>
+              </project>
+              """
+          )
+        );
+    }
+
 }

